File: LongestRepeatedSubstring/py2/lrs.py
# longest repeated substring soln in py2 for code eval by steven a dunn

import sys

def lcp(s1, s2):
	n = min(len(s1), len(s2))
	for i in range(n):
		if s1[i] != s2[i]:
			return s1[0:i]
	if s2[n-1:] == s1:
		return s1[0:n-1]
	return s1[0:n]

def find_lrs(line):
	suffixes = []
	for i in range(len(line)):
		suffixes.append(line[i:])
	suffixes = sorted(suffixes)
	lrs = ""
	for i in range(len(line) - 1):
		prefix = lcp(suffixes[i], suffixes[i+1])
		if len(prefix) > len(lrs):
			lrs = prefix
	return "".join(lrs.split())

f = open(sys.argv[1], 'r')
for line in f:
	print "".join(line.strip().split())
	result = find_lrs(line.strip())
	if not result:
		print "NONE"
	else:
		print result
f.close()

File: LongestRepeatedSubstring/cpp/LRS.cpp
// Longest Repeated Substring solution in C++ for CodeEval.com by Steven A. Dunn

#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using std::cout;
using std::endl;
using std::ifstream;
using std::min;
using std::remove_if;
using std::sort;
using std::string;
using std::vector;

string printLRS(string);
string findLRS(string);
string lcp(string, string);

int main (int argc, char* argv[])
{
  ifstream inputFile(argv[1]);
  string line;

  if (inputFile)
  {
    while (getline(inputFile, line))
    {
      cout << printLRS(line) << endl;
    }
    inputFile.close();
  }

  return 0;
}

string printLRS(string line)
{
  string result = findLRS(line);
  if (result.empty())
    return "NONE";
  else 
    return result;
}

string findLRS(string line)
{
  int len = line.size();
  vector<string> suffixes(len);
  for (int i = 0; i < len; ++i)
    suffixes[i] = line.substr(i, len);

  sort(suffixes.begin(), suffixes.end());

  string lrs = "";
  for (int i = 0; i < len - 1; ++i)
  {
    string prefix = lcp(suffixes[i], suffixes[i+1]);
    if(prefix.size() > lrs.size())
      lrs = prefix;
  }

  lrs.erase(remove_if(lrs.begin(), lrs.end(), ::isspace), lrs.end());
  return lrs;
}

string lcp(string suffix1, string suffix2)
{
  int n = min(suffix1.size(), suffix2.size());
  for (int i = 0; i < n; ++i)
  {
    if (suffix1[i] != suffix2[i])
      return suffix1.substr(0, i);
  }

  if (suffix2.substr(n-1) == suffix1 )
    return suffix1.substr(0, n - 1);

  return suffix1.substr(0, n);
}
